### ISSUE TITLE
Profiler's `InitializeCurrentThread` returns `CORPROF_E_UNSUPPORTED_CALL_SEQUENCE`

### ISSUE THREAD

**Issue by yegor-pelykh** [dotnet/runtime#103210]
### Description

I hooked certain Windows functions inside my DLL.

And I wanted to try to get the mixed mode call stack inside the hooked functions.

To do this, I made the profiler (`ICorProfilerCallback9`) work in parallel in the same DLL, which is initialized and works correctly.
Inside this profiler there is an ability to use profiler info (`ICorProfilerInfo8`) as is usually done.

The `ICorProfilerCallback9` and `ICorProfilerInfo8` instances are stored as a global variables.

In the intercepted function, I tried to call the `DoStackSnapshot` method through this global instance of the `ICorProfilerInfo8`.
These attempts were unsuccessful.

Next, I tried calling method `ICorProfilerInfo8::InitializeCurrentThread`, as I saw in some examples, in order to prepare the thread for calling profiler info methods.

But this `InitializeCurrentThread` method returns HRESULT `0x80131363` (`CORPROF_E_UNSUPPORTED_CALL_SEQUENCE`).

It seems to me that something is not working as expected.

Yes, I understand that I'm not calling the profiler info methods from the profiler callback functions, which is the correct way to do this. But it seems to me that it should be possible to use profiler info from other DLL functions and from other threads.

Could you please advise what exactly is working wrong? And how can this error be avoided?
And fix it if it can be fixed on your side.

Thank you!

### Reproduction Steps

-

### Expected behavior

I expect function `InitializeCurrentThread` to return `0` (`S_OK`) and correctly prepare the thread for the profiler to use on it.

### Actual behavior

`InitializeCurrentThread` method returns `0x80131363` (`CORPROF_E_UNSUPPORTED_CALL_SEQUENCE`).

**Comment by tommcdon** [comment]
@davmason @noahfalk

**Comment by davmason** [comment]
Hi @yegor-pelykh. What you're seeing is that some methods are marked as synchronous and can only be called when in a callback from the profiler APIs. `InitializeCurrentThread` is one of those methods, it will not allow you to call it unless it's from within a profiler callback or on a thread that has never ran managed code.

The good news is that if you are seeing this message it means that the thread is already set up - you should only need to call `InitializeCurrentThread` on a brand new thread that the runtime did not create.

**Comment by yegor-pelykh** [comment]
Hi @davmason,
Thanks for the clarification, I'll remove the call of `InitializeCurrentThread`.

But this solution still doesn't solve my main problem.
I started experimenting with `InitializeCurrentThread` because calling method `ICorProfilerInfo8::DoStackSnapshot` returns `0x80131360` (`CORPROF_E_STACKSNAPSHOT_UNSAFE`). This happens in the very first calls to this method, at the very early stage of the app work.
After the first 2 or 3 calls, further calls return `S_OK`.
For some reason I thought that the problem was somehow related to the thread not being initialized.
But if this is not related to this, could you please then tell me what it is related to?
And how can I make the call of the `ICorProfilerInfo8::DoStackSnapshot` succeed?

**Comment by davmason** [comment]
`DoStackSnapshot` will return `CORPROF_E_STACKSNAPSHOT_UNSAFE` until the runtime is fully initialized, that is what you are seeing. The profiler is loaded at startup very early on before the runtime is fully initialized and we prevent any calls until everything is set up and ready to run managed code.

You can either ignore the first few errors, or wait until the main method runs (by subscribing to JIT events and looking for the name).

**Comment by yegor-pelykh** [comment]
Could you please point me to the correct way to know when the main function is called? What callback should I use for this? What approximately should be used inside the callback to understand that this is the main function? I have no idea yet...

> You can either ignore the first few errors, or wait until the main method runs (by subscribing to JIT events and looking for the name).

Could you please point me to the correct way to know when the main method is called? I have no idea yet...
What callback should I use for this? What should be used inside the callback to understand that this is the main function?

**Comment by davmason** [comment]
There are two ways I can think of 

1. The way you described above. To be complete you would have to set `COR_PRF_MONITOR_JIT_COMPILATION` in your flags, then monitor both `JITCompilationFinished` and `JITCachedFunctionSearchFinished` for the Main method since it might be R2R compiled or jitted. You can use this code to grab the name: https://github.com/dotnet/runtime/blob/0686ce61ed1e1cb3cb420281a0154efa5d0d00d5/src/tests/profiler/native/profiler.cpp#L585-L663
2. It might be sufficient to just wait for any JIT event and start sampling. We do some managed helpers before Main though, so I am not 100% sure it would work but it probably will. If you set `COR_PRF_MONITOR_JIT_COMPILATION` and start sampling when you receive the first `JITCompilationFinished` callback it should be enough.

**Comment by yegor-pelykh** [comment]
Thank you very much for the code and detailed description!
I got over it.

There is also a problem with getting the stack trace, but I think this is not relevant to this topic, and I'll probably create a separate issue if I don't find a solution.

Feel free to close this issue.
